### ISSUE TITLE
Lighting/Texture issues

### ISSUE THREAD

**Issue by Dylan-at-LION** [rastapasta/react-native-gl-model-view#14]
Hello again. I figured I'd make an issue about two separate issues since they can be explained with the same example, but I'd be happy to split this into two issues if you'd prefer.

Here is a gist of the ModelAnimator component that I'm using which is basically just a stripped down copy of your Animations component in the example: https://gist.github.com/Dylan-at-LION/66b33abab4e222add280d90cf3858465

Here is a GIF demonstrating both issues, note that the GIF is heavily optimized and there are no dots on the sphere in actuality:

![orange](https://user-images.githubusercontent.com/22622898/40083650-ba1f6d8c-5849-11e8-8b76-8d0e265d733c.gif)

Issue 1: Lighting

When I rotate my model, the lighting changes. This doesn't seem be affected by the props `scale` or `translateZ` as I suspected, but it does tend to happen more with spherical models. Ideally there would be consistent 'global lighting' like in a studio or something. The orange.obj (linked below) is initialized with a scale of 3:

`
<ModelAnimator model={ 'orange.obj' }
                             texture={ 'orange2.jpg' }
                             scale={ 3 } />
`

Issue 2: Texturing

The texture files seem to be applying the entire image instead mapping UV coordinates. Here is the texture file the free model came with (from Turbosquid):

![color](https://user-images.githubusercontent.com/22622898/40084052-c11eb38a-584a-11e8-8f86-7272a26dfd03.jpg)

Here is a texture I made from scratch in Photoshop just to see if it would make any difference (no):

![orange2](https://user-images.githubusercontent.com/22622898/40084104-e3e35fec-584a-11e8-9e82-290ee0d41d68.jpg)

What I did notice is that Photoshop maps the texture in basically the same way onto the image, save for the top left part.

I am using strictly .obj files. The only texture I've gotten to work properly is the demon.model / demon.png from the example. I don't have any more .model files to test with yet but I'm trying to get my hands on one, to see if that has anything to do with it. I thought maybe the demon.png texture having the color black as its negative space had something to do with it, but when I made the background black on the orange texture, black simply showed up on the orange with the rest of it.

Here is the .obj of the orange: https://gist.github.com/Dylan-at-LION/aa45c6e16f17a49f993e2dec9f7314b5

The orange.obj in question has this at the top:
`mtllib Orange.mtl`

Could not providing the .mtl have anything to do with this?

Any help would be greatly appreciated even if only a point in the right direction. I don't really know Objective C but I'm not terribly intimidated by the prospect of poking around in there.

**Comment by Dylan-at-LION** [comment]
Cool, thanks for taking a look at it.

Here is the model on Turbosquid: https://www.turbosquid.com/FullPreview/Index.cfm/ID/756550

**Comment by PatriceVignola** [comment]
Ok, so far I can confirm that the problem only happens on iOS, as I thought:

![screenshot_2018-05-17-01-30-46](https://user-images.githubusercontent.com/5880386/40130670-ba098498-5972-11e8-90f3-ea43696966b5.png)
![screenshot_2018-05-17-01-32-53](https://user-images.githubusercontent.com/5880386/40130673-bc60b11c-5972-11e8-8b82-4ed67c33218b.png)

I spent quite a bit of time implementing the Android version and tested many .OBJ models, so this bug would have surprised me. On the other hand, I didn't really touch the iOS version, but I know the library we use has been deprecated for quite a while and isn't maintained anymore. Luckily, the library is not very large so if there's a problem with it we can easily make a fork and fix it, or use a more recent library. The latter option would probably be the preferred approach.

**Comment by Dylan-at-LION** [comment]
I see. Do you think if I made sense of the methods used in the Android version that would help debug GLView?: http://www.jpct.net/doc/com/threed/jpct/Texture.html#Texture(java.awt.Image)

I'm going to try making some tweaks to GLView I'm just not quite sure where to start. It may be a bit out of my depth but I'd really like to get this working on iOS.

Do you think the lighting issue is also at the GLView level?

Another idea I had was to try and convert an .obj to the .model format used in the example but I have no idea how to go about that yet, and it's hardly a long term solution.

**Comment by PatriceVignola** [comment]
I will dive into the problem as soon as I get the time, but in the meantime, you can start looking at [this file](https://github.com/nicklockwood/GLView/blob/master/GLView/Models/GLModel.m). Search for the string `vt` or the array `tempCoordData`. It fetches all the `vt` lines in the model file wnd add them to a list to process them after parsing the model.

Now that I think about it, the problem sounds weird because I successfully used this module to build a project with many models on iOS in the past and it successfully mapped the textures. One thint that's different with your model though is the fact that it uses 2 objects: the skin and the top/bottom (heart?) of the orange. It's possible that the parser only expecrs one object, therefore using the texture coordinates of the heart for the skin for example. You can spot different objects in the file by lines starting with the letter `o`, and check if the parser correctly manages them. This is where I will start when I get the chance to dive ibto the problem.

**Comment by Dylan-at-LION** [comment]
Thanks for the tips, looking into them now. Happy to hear that you'll dive into it when you get a chance.

**Comment by PatriceVignola** [comment]
Hey Dylan, I found the problem. For some reason, the y coordinate of the UV is inverted for .OBJ models, but not for .MODEL ones. I don't know exactly why yet since they both go through the same pipelines and they don't seem to be modified anywhere, but if I just multiply the y coordinate by -1, all .obj render correctly:

![orangeskin](https://user-images.githubusercontent.com/5880386/40194593-f5584024-5a45-11e8-818f-99df4ac932f7.png)
![smiley](https://user-images.githubusercontent.com/5880386/40194602-f9b477f0-5a45-11e8-9a38-55efc08dc4de.png)

Of course, with a sphere, you can't tell if the texture is mapped correctly or is upside down, so I tested with more complex objects like this horse:

![horse](https://user-images.githubusercontent.com/5880386/40194606-fef3d738-5a45-11e8-8aaf-f6b0910fc730.png)

I could only find a normal map for this object, but we can clearly see that the eyes, nose, muscles are at the right place, therefore the fix is working. If you want to test it on your machine, you can clone [GLView](https://github.com/nicklockwood/GLView) on your machine and add this line around **line 263 in GLModel.m**: `coords[1] *= -1;`. This will invert the UV y coordinate and your model will render correctly. Then, you can add a line to your Podfile that points to it.

Since the GLView repo is no longer maintained and the owner doesn't want to make bug fixes anymore, I did a fork and I'll fix the problem very soon. I don't want to push this fix just yet though because it feels like a hack and I want to understand why the coordinates are only being inverted for the .OBJ models.

**Comment by Dylan-at-LION** [comment]
Hi Patrice, that's great news! 

Does this look right to you?
![screen shot 2018-05-17 at 12 12 54 pm](https://user-images.githubusercontent.com/22622898/40198609-c58468de-59cb-11e8-9027-79c1152c9da3.png)

I also tried placing it here:
![screen shot 2018-05-17 at 12 44 53 pm](https://user-images.githubusercontent.com/22622898/40200155-71817f56-59d0-11e8-95b9-e2be7b0631a8.png)


I did a project "clean" and rebuilt everything but I don't think it worked for me cause something is messed up with my file structure.

I got GLView to compile by placing it here:

![screen shot 2018-05-17 at 12 11 37 pm](https://user-images.githubusercontent.com/22622898/40198873-82d03012-59cc-11e8-812d-904dc25caa6e.png)

I guess it's bad form but it worked for me for whatever reason, just wanted to see if I could get it working at all.

Whenever I make changes to RNGLModelView.m or any of the files on that level, just printing log messages really, they show up in the console just fine. But when I change anything in the GLView folder it doesn't seem to change anything (none of my messages make it to the console) unless it's an error and the build fails.

This is what my Podfile looks like: https://gist.github.com/Dylan-at-LION/3eb75752b714731c24d9126dbdb5116c

I have GLView installed as a Pod I just wasn't sure what to do with it as this is my first time using Cocoapods:
![screen shot 2018-05-17 at 12 27 11 pm](https://user-images.githubusercontent.com/22622898/40199253-c39cfeee-59cd-11e8-8009-445903c1a395.png)

Thanks again! Excited to get this working.

**Comment by Dylan-at-LION** [comment]
I "solved" the lighting issue by changing this value in GLLight.m:

`self.ambientColor = [UIColor colorWithWhite:**0.2** alpha:1.0];`

I changed 0.2 to 0.8 and now the ambient light is strong enough such that there aren't any dramatic shadows (at least not from simply rotating the model), which is basically what I was after.

**Comment by PatriceVignola** [comment]
Hey @Dylan-at-LION! Sorry for the huge delay, I was busy with other projects, but in the meantime I also made a [.model to .obj transcoder](https://www.npmjs.com/package/dot-model-2-obj?activeTab=readme) in order to diagnose the inverted texture problem and fix the consistency problem between the Android and iOS libraries.

Also, since 3D programs have the choice to export the textures in left-handed or right-handed coordinate systems, I added a `flipTexture` property to `RNGLModelView` to flip textures vertically. [I pushed it to the git repo](https://github.com/rastapasta/react-native-gl-model-view/commit/b1f1fd7ec364eed141e4bdfd2f96b481c80185f9), and it should be published to npm as soon as @rastapasta green lights it.

Feel free to try it out by temporarily replacing the dependency with the git repo if you want: `git+https://github.com/rastapasta/react-native-gl-model-view.git`.

Also, thanks for playing around with the light. I might push it as a fix until the light isn't attached to the model anymore.
